### PR TITLE
[6.0.x] Fix to #26451 - Temporal Table: Owned Entities support

### PR DESCRIPTION
**Pull request by maumar**
Added extensibility point to SharedTypeEntityExpandingExpressionVisitor so that we can create temporal table expressions representing owned entities, if their owner is also a temporal table expression.
Just like any other nav expansion, this only works for AsOf operation.

Also added internal interface to TableExpression, so that we can extract table metadata information (table name and schema) from provider specific table expressions.

### CODE REVIEW

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 1]
```diff
@@ -1375,6 +1398,23 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+                TableExpressionBase FindRootTableExpressionForColumn(TableExpressionBase table, string columnName)
+                {
+                    if (table is JoinExpressionBase joinExpressionBase)
+                    {
+                        return FindRootTableExpressionForColumn(joinExpressionBase.Table, columnName);
+                    }
+
+                    if (table is SelectExpression selectExpression)
+                    {
+                        var matchingProjection = (ColumnExpression)selectExpression.Projection.Where(p => p.Alias == columnName).Single().Expression;
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1411]: is this always guaranteed to be a column, given that it originally was a column? (binding of entity property)

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -1375,6 +1398,23 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+
+                        return FindRootTableExpressionForColumn(matchingProjection.Table, matchingProjection.Name);
+                    }
+
+                    return table;
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1416]: should this be provider specific also?

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 3]
```diff
@@ -10341,6 +10343,167 @@ public class CollectionViewModel25225
...
+
+            var subquery = context.MainEntities
+                .TemporalAsOf(date)
+                .Distinct()
+                .OrderByDescending(x => x.Id)
+                .Take(3);
+
+            var subqueryResult = async ? await subquery.ToListAsync() : subquery.ToList();
+
+            var complex1 = context.MainEntities.TemporalAsOf(date)
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10374]: nested select

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 4]
```diff
@@ -10341,6 +10343,167 @@ public class CollectionViewModel25225
...
+                .Join(context.MainEntities, x => x.Id, x => x.Id, (o, i) => new { o, i })
+                .Distinct().OrderByDescending(x => x.o.Id).Take(3)
+                .Join(context.MainEntities, xx => xx.o.Id, x => x.Id, (o, i) => new { o, i });
+
+            var complex1Result = async ? await complex1.ToListAsync() : complex1.ToList();
+
+            var complex2 = context.MainEntities
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10381]: extracting source table based on non-trivial alias ([t0].[Id0] -> [t].[Id0] -> [m0].[Id] AS [Id0])

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 5]
```diff
@@ -1375,6 +1398,23 @@ protected override Expression VisitExtension(Expression extensionExpression)
                 }
 
                 return innerShaper;
+
+                TableExpressionBase FindRootTableExpressionForColumn(TableExpressionBase table, string columnName)
+                {
+                    if (table is JoinExpressionBase joinExpressionBase)
+                    {
+                        return FindRootTableExpressionForColumn(joinExpressionBase.Table, columnName);
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1406]: avoid recursion and assign this to table itself.

File: src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        /// <summary>
+        /// Sql expression factory.
+        /// </summary>
+        public ISqlExpressionFactory SqlExpressionFactory { get; init; }
+
+        /// <summary>
+        ///     Creates a new instance of the <see cref="SharedTypeEntityExpansionHelper" /> class.
+        /// </summary>
+        /// <param name="sqlExpressionFactory">Sql expression factory.</param>
+        public SharedTypeEntityExpansionHelper(ISqlExpressionFactory sqlExpressionFactory)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs:21]: Dependencies object

File: src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    /// <inheritdoc/>
+    public class SharedTypeEntityExpansionHelper : ISharedTypeEntityExpansionHelper
+    {
+        /// <summary>
+        /// Sql expression factory.
+        /// </summary>
+        public ISqlExpressionFactory SqlExpressionFactory { get; init; }
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs:15]: expose dependencies object not the service

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs (hunk 3 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10374]: split each in diff test.

Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10374]: Create a new test base deriving from NonSharedModelTestBase with temporal traits.

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 8]
```diff
@@ -10341,6 +10343,167 @@ public class CollectionViewModel25225
...
+    ORDER BY [t].[Id] DESC
+) AS [t0]
+INNER JOIN [MainEntity] AS [m1] ON [t0].[Id] = [m1].[Id]
+LEFT JOIN [OwnedEntity] AS [o] ON [t0].[Id] = [o].[MainEntity26451Id]
+LEFT JOIN [OwnedEntity] FOR SYSTEM_TIME AS OF '2000-01-01T00:00:00.0000000' AS [o0] ON [t0].[Id0] = [o0].[MainEntity26451Id]
+LEFT JOIN [OwnedEntity] AS [o1] ON [m1].[Id] = [o1].[MainEntity26451Id]
+ORDER BY [t0].[Id] DESC");
+        }
+
+        [ConditionalFact]
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:10450]: Conditional theory. Verify exception for sync/async both.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 9]
```diff
@@ -1375,6 +1398,34 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+                        }
+                        else if (table is SelectExpression selectExpression)
+                        {
+                            var matchingProjection = (ColumnExpression)selectExpression.Projection.Where(p => p.Alias == columnName).Single().Expression;
+                            table = matchingProjection.Table;
+                            columnName = matchingProjection.Name;
+                        }
+                        else if (table is SetOperationBase setOperationBase)
+                        {
+                            table = setOperationBase.Source1;
```
Comment (maumar) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1419]: this *should* be ok, since for set ops, temporal operation must be identical for both sources

File: src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    /// <summary>
+    ///     Service which helps with various aspects of shared type entity expansion extensibility.
+    /// </summary>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-providers">Implementation of database providers and extensions</see>
+    ///     and <see href="https://aka.ms/efcore-how-queries-work">How EF Core queries work</see> for more information.
+    /// </remarks>
+    public interface ISharedTypeEntityExpansionHelper
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs:16]: service characteristic is missing in doc.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 9 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1419]: Do not really like while loop either. :trollface: 

Idea is that if table SelectExpression we need to go nested inside so we use recursion for that.
But other 2 types are just unwrapping a table/select expression inside something different. So we do this unwrapping first.
code would be like
```C#
FindRootTableExpressionForColumn(...)
{
if (table is JoinExpressionBase)
{
    table = joinExpressionBase.Table;
}
else if (table is SetOperationBase)
{
    table = setOperationBase.Source1;
}

if (table is SelectExpression)
{
    var matchingProjection = ...
    return FindRootTableExpressionForColumn(....)
}

return table;
```

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1427]: in what sense?

At this point I think, given that we have to find on deeper level, we should have implementation temporal times as annotations over table expression

File: src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+        /// <summary>
+        ///     Dependencies for this service.
+        /// </summary>
+        protected virtual SharedTypeEntityExpansionHelperDependencies Dependencies { get; }
+
+        /// <summary>
+        ///     Creates a new instance of the <see cref="SharedTypeEntityExpansionHelper" /> class.
+        /// </summary>
+        /// <param name="dependencies">Dependencies for this service.</param>
+        public SharedTypeEntityExpansionHelper(SharedTypeEntityExpansionHelperDependencies dependencies)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SharedTypeEntityExpansionHelper.cs:21]: ctor before protected properties.

File: src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        /// <inheritdoc/>
+        public override SelectExpression CreateInnerSelectExpression(
+            TableExpressionBase sourceTable,
+            IEntityType targetEntityType)
+        {
+            if (sourceTable is TemporalTableExpression)
+            {
+                if (sourceTable is TemporalAsOfTableExpression temporalAsOf)
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs:34]: You could avoid 2 checks for AsOf by moving this whole if block outside at top. And then another if for exception.

File: test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs [hunk 13]
```diff
@@ -12,6 +12,7 @@
 using System.Linq.Expressions;
 using System.Reflection;
 using System.Text;
+using System.Text.Json;
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/QueryBugsTest.cs:15]: ???

File: test/EFCore.SqlServer.FunctionalTests/Query/TemporalOwnedEntityMappedToDifferentTableSqlServerTest.cs [hunk 14]
```diff
@@ -0,0 +1,268 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore.SqlServer.Internal;
+using Microsoft.EntityFrameworkCore.TestUtilities;
+using Xunit;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    [SqlServerCondition(SqlServerCondition.SupportsTemporalTablesCascadeDelete)]
+    public class TemporalOwnedEntityMappedToDifferentTableSqlServerTest : NonSharedModelTestBase
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/TemporalOwnedEntityMappedToDifferentTableSqlServerTest.cs:14]: TemporalTableSqlServerTest
which can have all customer models regarding temporal table feature. We can fork in owned and regular if we need to in future.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 2 above)
Comment (maumar) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1427]: im thinking of some custom sql-server specific structure which encapsulates Table expression (similar to SelectExpression) - this would require provider specific logic to extract the correct table information, since relational wouldn't know anything about the custom expression type.

Annotation seems bit overkill, we only need this information in this one specific place (so far)

Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1420]: we don't allow custom table sources in that sense. But we should investigate more how we want to expand support for it. Can you file an issue?

File: src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    ///     <see cref="DbContext" /> instance will use its own instance of this service.
+    ///     The implementation may depend on other services registered with any lifetime.
+    ///     The implementation does not need to be thread-safe.
+    /// </para>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-providers">Implementation of database providers and extensions</see>
+    ///     and <see href="https://aka.ms/efcore-how-queries-work">How EF Core queries work</see> for more information.
+    /// </remarks>
+    public interface ISharedTypeEntityExpansionHelper
+    {
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs:24]: This should be prefixed with relational.

File: src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+                var table = targetEntityType.GetTableMappings().Single().Table;
+                var temporalTableExpression = new TemporalAsOfTableExpression(table, temporalAsOf.PointInTime);
+
+                return Dependencies.SqlExpressionFactory.Select(targetEntityType, temporalTableExpression);
+            }
+
+            if (sourceTable is TemporalTableExpression)
+            {
+                throw new InvalidOperationException(
+                    SqlServerStrings.TemporalOwnedTypeMappedToDifferentTableOnlySupportedForAsOf("AsOf"));
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs:43]: If the exception message contains AsOf in the name why do we need  to pass it as arg too?

File: test/EFCore.SqlServer.FunctionalTests/Query/TemporalTableSqlServerTest.cs [hunk 17]
```diff
@@ -0,0 +1,310 @@
...
+                        {
+                            return context.MainEntities.TemporalAll().ToListAsync();
+                        }
+                        else
+                        {
+                            var result = context.MainEntities.TemporalAll().ToList();
+
+                            return Task.FromResult(result);
+                        }
+                    })).Message;
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/Query/TemporalTableSqlServerTest.cs:259]:
```
var message = async
    ? (await Assert.ThrowsAsync<>(() => abc.ToListAsync())).Message
    : Assert.Throws<>(() => abc.ToList()).Message;
```

File: src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs (hunk 16 above)
Comment (maumar) [reply src/EFCore.SqlServer/Query/Internal/SqlServerSharedTypeEntityExpansionHelper.cs:43]: @AndriySvyryd suggestion when reviewing original temp tables work, see TemporalNavigationExpansionOnlySupportedForAsOf

File: src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs (hunk 15 above)
Comment (maumar) [reply src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs:24]: do we have clear rules about it? seems inconsistent from what I saw - we always add prefix for stuff that in in core, but for interfaces that are relational only, we sometimes add the prefix and sometimes we don't

Comment (smitpatel) [reply src/EFCore.Relational/Query/ISharedTypeEntityExpansionHelper.cs:24]: In query we do, afaik.

**Comment by maumar** [comment]
updated

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1415]: Should be fine.

**Comment by smitpatel** [comment]
LGTM

**Review by xxyuze — APPROVED** [review]

**Comment by maumar** [comment]
@smitpatel new version up, @ajcvickers for patch evaluation

File: src/EFCore.Relational/Query/IRelationalSharedTypeEntityExpansionHelper.cs [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    ///     The service lifetime is <see cref="ServiceLifetime.Scoped" />. This means that each
+    ///     <see cref="DbContext" /> instance will use its own instance of this service.
+    ///     The implementation may depend on other services registered with any lifetime.
+    ///     The implementation does not need to be thread-safe.
+    /// </para>
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-providers">Implementation of database providers and extensions</see>
+    ///     and <see href="https://aka.ms/efcore-how-queries-work">How EF Core queries work</see> for more information.
+    /// </remarks>
+    public interface IRelationalSharedTypeEntityExpansionHelper
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/IRelationalSharedTypeEntityExpansionHelper.cs:23]: If there's no way to avoid adding API for this fix, I'd prefer that the interface as made pubternal for the patch and all references marked with Internal attribute

File: src/EFCore.Relational/Query/Internal/IRelationalSharedTypeEntityExpansionHelper.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+    /// <remarks>
+    ///     See <see href="https://aka.ms/efcore-docs-providers">Implementation of database providers and extensions</see>
+    ///     and <see href="https://aka.ms/efcore-how-queries-work">How EF Core queries work</see> for more information.
+    /// </remarks>
+    public interface IRelationalSharedTypeEntityExpansionHelper
+    {
+        /// <summary>
+        /// Creates a <see cref="TableExpressionBase" /> related to the source table for a given entity type.
+        /// </summary>
+        public TableExpressionBase CreateRelatedTableExpression(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/IRelationalSharedTypeEntityExpansionHelper.cs:28]: internal docs

File: src/EFCore.Relational/Query/Internal/ITableMetadata.cs [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+    /// </summary>
+    public interface ITableMetadata
+    {
+        /// <summary>
+        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        string Name { get; }
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/Internal/ITableMetadata.cs:20]: Instead add ITable property.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 21]
```diff
@@ -1255,11 +1265,33 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
                 var foreignKey = navigation.ForeignKey;
                 if (navigation.IsCollection)
                 {
-                    var innerShapedQuery = CreateShapedQueryExpression(
-                        targetEntityType, _sqlExpressionFactory.Select(targetEntityType));
+                    var innerShapedQuery = default(ShapedQueryExpression);
+                    if (!useOldBehavior)
+                    {
+                        var innerSelectExpression = BuildInnerSelectExpressionForOwnedTypeMappedToDifferentTable(
+                        entityProjectionExpression,
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1283]: nit: indent

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs [hunk 22]
```diff
@@ -1255,11 +1265,33 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+                var useOldBehavior = AppContext.TryGetSwitch("Microsoft.EntityFrameworkCore.Issue26469", out var enabled)
+                    && enabled;
+
                 var foreignKey = navigation.ForeignKey;
                 if (navigation.IsCollection)
                 {
-                    var innerShapedQuery = CreateShapedQueryExpression(
-                        targetEntityType, _sqlExpressionFactory.Select(targetEntityType));
+                    var innerShapedQuery = default(ShapedQueryExpression);
+                    if (!useOldBehavior)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitor.cs:1280]: can if on innerSelectExpression construction. Method to create ShapedQuery afterwards is same call.

File: src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitorDependencies.cs [hunk 23]
```diff
@@ -72,5 +76,10 @@ public sealed record RelationalQueryableMethodTranslatingExpressionVisitorDepend
         ///     The SQL expression factory.
         /// </summary>
         public ISqlExpressionFactory SqlExpressionFactory { get; init; }
+
+        /// <summary>
+        ///     Shared type entity expansion helper.
+        /// </summary>
+        public IRelationalSharedTypeEntityExpansionHelper RelationalSharedTypeEntityExpansionHelper { get; init; }
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalQueryableMethodTranslatingExpressionVisitorDependencies.cs:83]: should annotate with internal

File: src/EFCore.Relational/Query/RelationalSharedTypeEntityExpansionHelper.cs [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+using System.Linq;
+using Microsoft.EntityFrameworkCore.Infrastructure;
+using Microsoft.EntityFrameworkCore.Metadata;
+using Microsoft.EntityFrameworkCore.Query.Internal;
+using Microsoft.EntityFrameworkCore.Query.SqlExpressions;
+
+namespace Microsoft.EntityFrameworkCore.Query
+{
+    /// <inheritdoc/>
+    public class RelationalSharedTypeEntityExpansionHelper : IRelationalSharedTypeEntityExpansionHelper
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/RelationalSharedTypeEntityExpansionHelper.cs:13]: internal

File: src/EFCore.Relational/Query/SqlExpressionFactoryDependencies.cs [hunk 25]
```diff
@@ -54,20 +55,30 @@ public sealed record SqlExpressionFactoryDependencies
         ///     </para>
         /// </summary>
         [EntityFrameworkInternal]
-        public SqlExpressionFactoryDependencies(IModel model, IRelationalTypeMappingSource typeMappingSource)
+        public SqlExpressionFactoryDependencies(
+            IModel model,
+            IRelationalTypeMappingSource typeMappingSource,
+            IRelationalSharedTypeEntityExpansionHelper relationalSharedTypeEntityExpansionHelper)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressionFactoryDependencies.cs:61]: why is this needed?

**Comment by smitpatel** [comment]
All the new API which is kinda temporary for the patch, mark them with internal
If classes put in internal namespace, if properties put attribute. Docs on all of them to internal template.

File: src/EFCore.Relational/Query/SqlExpressionFactoryDependencies.cs (hunk 25 above)
Comment (maumar) [reply src/EFCore.Relational/Query/SqlExpressionFactoryDependencies.cs:61]: remnant from the previous version - need to fix non-owned shared type, removed

File: src/EFCore.Relational/Query/Internal/ITableMetadata.cs (hunk 20 above)
Comment (maumar) [reply src/EFCore.Relational/Query/Internal/ITableMetadata.cs:20]: we already have Name and schema on TableExpression, so wanted to mirror that

**Comment by ajcvickers** [comment]
Note from triage: holding off on patch for now.

**Comment by maumar** [comment]
closing this, will issue a different PR soon
